core-bundle/src/Pow/Altcha/Altcha.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Pow\Altcha;
+
+use Contao\CoreBundle\Entity\PowAltcha;
+use Contao\CoreBundle\Pow\Altcha\Config\AlgorithmConfig;
+use Contao\CoreBundle\Pow\Altcha\Exception\InvalidAlgorithmException;
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Exception;
+use Doctrine\DBAL\Types\Types;
+use Doctrine\ORM\EntityManagerInterface;
+
+class Altcha
+{
+    public function __construct(
+        private readonly EntityManagerInterface $entityManager,
+        private readonly Connection $connection,
+        private readonly string $secret,
+        private readonly string $altchaAlgorithm,
+        private readonly int $altchaRangeMin,
+        private readonly int $altchaRangeMax,
+        private readonly int $altchaChallengeExpiry,
+    ) {
+    }
+
+    /**
+     * @throws InvalidAlgorithmException
+     * @throws Exception
+     */
+    public function createChallenge(string|null $salt = null, int|null $number = null): AltchaChallenge
+    {
+        $salt = $salt ?? bin2hex(random_bytes(12));
+        $number = $number ?? random_int($this->altchaRangeMin, $this->altchaRangeMax);
+
+        if (!\in_array($this->altchaAlgorithm, AlgorithmConfig::ALGORITHM_ALL, true)) {
+            $strError = 'Invalid algorithm selected. It has to be set to one of these "%s".';
+
+            throw new InvalidAlgorithmException(sprintf($strError, implode('", "', AlgorithmConfig::ALGORITHM_ALL)));
+        }
+
+        $algorithm = str_replace('-', '', strtolower($this->altchaAlgorithm));
+
+        // Generate the challenge
+        $challenge = hash($algorithm, $salt.$number);
+
+        // Generate the signature
+        $signature = hash_hmac($algorithm, $challenge, $this->secret);
+
+        // The challenge expires in 1 hour (default) and it must be saved in the database
+        // to prevent replay attacks.
+        $expiryDate = date('Y-m-d\\TH:i:sP', time() + $this->altchaChallengeExpiry);
+        $objPowAltcha = new PowAltcha($challenge, new \DateTimeImmutable($expiryDate));
+
+        // Insert record into tl_pow_altcha
+        $this->entityManager->persist($objPowAltcha);
+        $this->entityManager->flush();
+
+        return new AltchaChallenge($this->altchaAlgorithm, $challenge, $salt, $signature);
+    }
+
+    public function isValidPayload(string $payload): bool
+    {
+        $json = json_decode(base64_decode($payload, true), true);
+
+        if (null === $json) {
+            return false;
+        }
+
+        if ($this->isReplay($json)) {
+            return false;
+        }
+
+        $rowsAffected = $this->connection->executeStatement(
```

Comment:
This should go to the repository.

Reply:
@Toflar What do you mean with:
> We should put this in the repository.

Reply:
You've created a Doctrine Entity and you are using the entity manager to persist that entity. Querying should happen in the dedicated Doctrine repository for that entity :) Like e.g. our `CronJobRepository`.